config/management/src/smoke_test.rs

```diff
@@ -12,59 +12,73 @@ use libra_secure_storage::{CryptoStorage, KVStorage, Value};
+        Self {
+            storage_helper: StorageHelper::new(),
+            num_validators,
+            swarm_path,
+        }
     }
-}
 
-#[test]
-fn smoke_test() {
-    LibraNode::prepare();
-    let helper = StorageHelper::new();
-    let num_validators = 5;
-    let shared = "_shared";
-    let association = "association";
-    let association_shared = association.to_string() + shared;
-
-    // Step 1) Prepare the layout
-    let mut layout = Layout::default();
-    layout.association = vec![association_shared.to_string()];
-    layout.operators = (0..num_validators)
-        .map(|v| (v.to_string() + shared))
-        .collect();
-
-    let mut common_storage = helper.storage(crate::constants::COMMON_NS.into());
-    let layout_value = Value::String(layout.to_toml().unwrap());
-    common_storage
-        .set(crate::constants::LAYOUT, layout_value)
-        .unwrap();
-
-    // Step 2) Set association key
-    helper.initialize(association.into());
-    helper
-        .association_key(&association, &association_shared)
-        .unwrap();
-
-    // Step 3) Prepare validators
-    let temppath = TempPath::new();
-    temppath.create_as_dir().unwrap();
-    let swarm_path = temppath.path().to_path_buf();
-
-    let mut configs = Vec::new();
-    for i in 0..num_validators {
-        let ns = i.to_string();
-        let ns_shared = ns.clone() + shared;
-        helper.initialize(ns.clone());
-
-        let operator_key = helper.operator_key(&ns, &ns_shared).unwrap();
+    fn secure_backend(&self, ns: &str, usage: &str) -> SecureBackend {
+        let original = self.storage_helper.path();
+        let dst_base = self.swarm_path.path();
+        let mut dst = dst_base.to_path_buf();
+        dst.push(format!("{}_{}", usage, ns));
+        std::fs::copy(original, &dst).unwrap();
+
+        let mut storage_config = OnDiskStorageConfig::default();
+        storage_config.path = dst;
+        storage_config.set_data_dir(PathBuf::from(""));
+        storage_config.namespace = Some(ns.into());
+        SecureBackend::OnDiskStorage(storage_config)
+    }
+
+    fn create_layout(&self) {
+        let mut layout = Layout::default();
+        layout.association = vec![ASSOCIATION_SHARED.into()];
+        layout.operators = (0..self.num_validators)
+            .map(|v| (v.to_string() + SHARED))
+            .collect();
+
+        let mut common_storage = self
+            .storage_helper
+            .storage(crate::constants::COMMON_NS.into());
+        let layout_value = Value::String(layout.to_toml().unwrap());
+        common_storage
+            .set(crate::constants::LAYOUT, layout_value)
+            .unwrap();
+    }
+
+    fn create_association(&self) {
+        self.storage_helper.initialize(ASSOCIATION.into());
+        self.storage_helper
+            .association_key(ASSOCIATION, ASSOCIATION_SHARED)
+            .unwrap();
+    }
+
+    fn initialize_validator_config(&self, index: usize) -> NodeConfig {
```

Comment:
One of the disadvantages of this refactoring is that it becomes more difficult to identity the distinct steps that need to happen to bootstrap the network and produce genesis (e.g., the steps outlined in the smoke test as comments). Perhaps we could try and keep some of the steps by: (i) adding a comment to the functions "initialize_validator_config" and "finish_validator_config", to explain what each functions does; and/or (ii) renaming the functions?

For example, the method "initialize_validator_config" internally uploads the validator operator key to the shared repo, creates a local NodeConfig for the validator, and then uploads the validator config to the shared repo. Likewise, "finish_validator_config" internally constructs genesis, creates and inserts a waypoint, verifies genesis and then returns a NodeConfig for the validator.

Note: for me, I think the confusing comes from the fact that the term "validator config" is overloaded; here it can mean either: (i) a configuration file, such as NodeConfig; (ii) the name of the command executed by each operator to upload a ValidatorConfig transaction to the shared repo; or (iii) the ValidatorConfig transaction itself.

Reply:
I think I followed the steps in the original tests for each function i.e. every function corresponds to one step in the `build_swarm`